### PR TITLE
Allow plugins to override toEqual types with plugin.smartEqRules

### PR DESCRIPTION
**Pull request by hasparus**
Goddamn, this was pretty hard.

## What's changed?

1. I've added `SmartEqRules` type — a registry of smart equality rules added by plugins. `Expectation.toEqual` parameter type is derived from it.

2. I needed a helper type to work around the case where registering the rule of type `SmartEqRule<unknown, unknown>` broke whole type of `toEqual`. This is `createPlugin.SmartEqRulesOf`.

3. That was a great opportunity to rework how the plugins register their types, so I've added:
   - `createPlugin` constrained identity function, as previously one could not use `PluginConfig` without erasing types of their plugin
   - helper types `createPlugin.MatchersOf`, `createPlugin.Validators`, and the aforementioned `createPlugin.SmartEqRulesOf`.
  
   This is how it looks like.
   
   ```ts
   const plugin = createPlugin({
     // your matchers, validators, and smartEqRules
   })
   
   declare module 'earljs' {
     interface Expect extends createPlugin.MatchersOf<typeof plugin> {}
     interface Expectation<_> extends createPlugin.ValidatorsOf<typeof plugin> {}
     interface SmartEqRules extends createPlugin.SmartEqRulesOf<typeof plugin> {}
   }
   ```
   
4. I've added tests to cover this feature to both `example-plugin/test` and `earljs/test`.

5. `plugin.smartEqRules` can now be a dictionary, and unfortunately it must be, as we can only declaration merge interfaces.

   We could treat arrays as `Record<number, Rule>`, but whenever user adds two plugins, we'd land on `ts(2717)`.
   
   As a workaround, we could improve `SmartEqRulesOf` or provide a new utility type, something similar to the following:
   
```ts
type SmartEqRulesOfArray<TPlugin extends PluginConfig & { smartEqRules: [...any[]] }, TPluginName extends string> = {
  [I in Exclude<keyof TPlugin['smartEqRules'], keyof any[]> as `${TPluginName} : ${I &
    string}`]: TPlugin['smartEqRules'][I]
}

// type _ = {
//     "earljs-ethers@1.0.0 : 0": SmartEqRule<number, string>;
//     "earljs-ethers@1.0.0 : 1": SmartEqRule<string, number>;
// }
type _ = SmartEqRulesOfArray<
  {
    smartEqRules: [SmartEqRule<number, string>, SmartEqRule<string, number>]
  },
  'earljs-ethers@1.0.0'
>
```

### CODE REVIEW

File: packages/earljs/src/validators/smartEq.ts [hunk 1]
```diff
@@ -99,8 +99,18 @@ export function smartEq(actual: any, expected: any, strict: boolean = true, seen
...
 }
+
+export interface SmartEqRules {}
+
+export declare namespace SmartEqRules {
+  export type Expected<TEqRule> = TEqRule extends SmartEqRule<any, infer TExpected> ? TExpected : never
+}
+
+export type ExpectedEqual<TActual> =
+  | TActual
```
Comment (krzkaczor) [inline packages/earljs/src/validators/smartEq.ts:115]: This type needs to be recusive.

**Review by krzkaczor — COMMENTED** [review]
`expect({ nested: new Set([1, 2, 3]) }).toEqual({ nested: [1, 2, 3] })`

File: packages/earljs/src/validators/smartEq.ts (hunk 1 above)
Comment (hasparus) [reply packages/earljs/src/validators/smartEq.ts:116]: Nice catch! Should be working now.

**Review by krzkaczor — APPROVED** [review]
LFG!

**Pull request merged**